handler/src/main/java/io/netty/handler/ssl/TlsMasterKeyHandler.java

```diff
@@ -0,0 +1,105 @@
+import javax.net.ssl.SSLSession;
+import java.lang.reflect.Field;
+
+/**
+ * The {@link TlsMasterKeyHandler} is a channel-handler you can include in your pipeline to consume the master key
+ * & session identifier for a TLS session.
+ * This can be very useful when combined with {@link TlsMasterKeyBiConsumer.WiresharkTlsMasterKeyBiConsumer} as it will
+ * log the secret & identifier in a format that is consumable by Wireshark -- allowing easy decryption of pcap/tcpdumps.
+ */
+public class TlsMasterKeyHandler extends ChannelInboundHandlerAdapter {
+
+    /**
+     * The JRE SSLSessionImp cannot be imported
+     */
+    private static final Class<?> SSL_SESSIONIMPL_CLASS;
+
+    /**
+     * A system property that can be used to turn on/off the {@link TlsMasterKeyHandler} dynamically without having
+     * to edit your pipeline.
+     * <code>-Dio.netty.tls.masterKeyHandler=true</code>
+     */
+    private static final String SYSTEM_PROP_KEY = "io.netty.tls.masterKeyHandler";
+
+    static {
+        try {
+            SSL_SESSIONIMPL_CLASS = Class.forName("sun.security.ssl.SSLSessionImpl");
+        } catch (ClassNotFoundException e) {
+            throw new Error("Could not load SSLSessionImpl");
+        }
+    }
+
+    private final TlsMasterKeyBiConsumer consumer;
+
+    /**
+     * Constructor. Provide a consumer to your liking.
+     * @param consumer
+     */
+    public TlsMasterKeyHandler(TlsMasterKeyBiConsumer consumer) {
+        this.consumer = consumer;
+    }
+
+    /**
+     * Constructor. Will set the consumer to one that does nothing.
+     */
+    public TlsMasterKeyHandler() {
+        this(TlsMasterKeyBiConsumer.DoNothingTlsMasterKeyBiConsumer.INSTANCE);
+    }
+
+    @Override
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) {
+
+        //only try to log the session info if the ssl handshake has successfully completed.
+        if (evt instanceof SslHandshakeCompletionEvent && evt == SslHandshakeCompletionEvent.SUCCESS) {
+
+            boolean shouldHandle = Boolean.getBoolean(SYSTEM_PROP_KEY);
+
+            //if the handler is not an instance of DoNothingTlsMasterKeyBiConsumer try to handle it
+            if (shouldHandle && TlsMasterKeyBiConsumer.DoNothingTlsMasterKeyBiConsumer.INSTANCE != consumer) {
+                final SslHandler handler = ctx.pipeline().get(SslHandler.class);
+                final SSLEngine engine = handler.engine();
+                final SSLSession sslSession = engine.getSession();
+
+                //the OracleJVM does not expose a way to get the master secret, so try to use reflection to get it.
+                try {
+                    if (sslSession.getClass().equals(SSL_SESSIONIMPL_CLASS)) {
+                        Field masterSecretField = SSL_SESSIONIMPL_CLASS.getDeclaredField("masterSecret");
+                        masterSecretField.setAccessible(true);
+                        SecretKey secretKey = (SecretKey) masterSecretField.get(sslSession);
```

Comment:
Should we use `PlatformDependent` (Which will use usafe) to access it ? This way it will also work with jigsaw.

Reply:
I'm not familiar with this class: do I just need to access the fieldOffset then get the object ?
(no need to set Accessible on the field?)

Reply:
Sorry I meant to say use `ReflectionUtil` 🤦‍♂️